File: src/main/scala/jk_5/nailed/server/mappack/metadata/xml/XmlMappackTeam.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.mappack.metadata.xml

import jk_5.nailed.api.chat.ChatColor
import jk_5.nailed.api.mappack.{MappackConfigurationException, MappackTeam}
import org.jdom2.Element

/**
 * No description given
 *
 * @author jk-5
 */
class XmlMappackTeam(element: Element) extends MappackTeam {
  if(element.getChild("id", element.getNamespace) == null) throw new MappackConfigurationException("Invalid team. Team doesn't have an id element")
  if(element.getChild("name", element.getNamespace) == null) throw new MappackConfigurationException("Invalid team. Team doesn't have a name element")
  if(element.getChild("color", element.getNamespace) == null) throw new MappackConfigurationException("Invalid team. Team doesn't have a color")

  override val id = element.getChild("id", element.getNamespace).getText
  override val name = element.getChild("name", element.getNamespace).getText
  override val color = ChatColor.getByName(element.getChild("color", element.getNamespace).getText)

  if(color == null) throw new MappackConfigurationException("Invalid team color '" + element.getChild("color", element.getNamespace).getText + "'. Color does not exist")
}

File: src/main/java/jk_5/nailed/server/mappack/metadata/xml/XmlMappackTeam.java
package jk_5.nailed.server.mappack.metadata.xml;

import jk_5.nailed.api.chat.ChatColor;
import jk_5.nailed.api.mappack.MappackConfigurationException;
import jk_5.nailed.api.mappack.metadata.MappackTeam;
import jk_5.nailed.api.scoreboard.Visibility;
import org.jdom2.Element;

public class XmlMappackTeam implements MappackTeam {

    private final String id;
    private final String name;
    private final ChatColor color;
    private final boolean friendlyFire;
    private final boolean friendlyInvisiblesInvisible;
    private final Visibility nameTagVisibility;
    private final Visibility deathMessageVisibility;

    public XmlMappackTeam(Element element) throws MappackConfigurationException {
        if(element.getChild("id", element.getNamespace()) == null){
            throw new MappackConfigurationException("Invalid team. Team doesn't have an id element");
        }
        if(element.getChild("name", element.getNamespace()) == null){
            throw new MappackConfigurationException("Invalid team. Team doesn't have a name element");
        }
        if(element.getChild("color", element.getNamespace()) == null){
            throw new MappackConfigurationException("Invalid team. Team doesn't have a color");
        }

        this.id = XmlUtils.getRequiredText(element, "id");
        this.name = XmlUtils.getRequiredText(element, "name");
        this.color = XmlUtils.getRequiredChatColor(element, "color");
        this.friendlyFire = XmlUtils.getBoolean(element, "friendlyFire", true);
        this.friendlyInvisiblesInvisible = XmlUtils.getBoolean(element, "friendlyInvisiblesInvisible", true);
        this.nameTagVisibility = XmlUtils.getVisibility(element, "nameTagVisibility", Visibility.ALWAYS);
        this.deathMessageVisibility = XmlUtils.getVisibility(element, "deathMessageVisibility", Visibility.ALWAYS);
    }

    @Override
    public String id() {
        return this.id;
    }

    @Override
    public String name() {
        return this.name;
    }

    @Override
    public ChatColor color() {
        return this.color;
    }

    @Override
    public boolean isFriendlyFire() {
        return this.friendlyFire;
    }

    @Override
    public boolean areFriendlyInvisiblesInvisible() {
        return this.friendlyInvisiblesInvisible;
    }

    @Override
    public Visibility getNameTagVisibility() {
        return this.nameTagVisibility;
    }

    @Override
    public Visibility getDeathMessageVisibility() {
        return this.deathMessageVisibility;
    }
}
